std/algorithm/iteration.d

```diff
@@ -905,182 +870,91 @@ iteration).
 
 Params:
     fun = function to apply to _each element of the range
-    r = range or iterable over which `each` iterates
+    range = range or iterable over which `each` iterates
 
 Returns: `Yes.each` if the entire range was iterated, `No.each` in case of early
 stopping.
 
 See_Also: $(REF tee, std,range)
  */
-template each(alias fun = "a")
+public Flag!"each" each(alias fun = "a", Range)(auto ref Range range)
+if (isInputRange!Range || isStaticArray!Range || hasMember!(Range, "opApply"))
```

Comment:
whew, nice, it can be reduced to a single function

Reply:
I don't think this is something that should be celebrated.

```D
        static if (is(typeof(f1!()))) return f1();
        else static if (is(typeof(f2!()))) return f2();
        else static if (is(typeof(f3!()))) return f3();
        else static if (is(typeof(f4!()))) return f4();
        else static if (is(typeof(f5!()))) return f5();
        else static if (is(typeof(f6!()))) return f6();
        else static if (is(typeof(f7!()))) return f7();
        else static assert(0, "Cannot iterate " ~ Range.stringof ~ " with each().");
```

Reply:
The "celebration" was in regard to moving `each` from a member in a template to a top-level function.

I agree that the "try each pattern and see what works" strategy does raise an eyebrow, and am always on lookout for better solutions - please advise. I took inspiration from @pbackus's amazing [sumtype library](https://code.dlang.org/packages/sumtype) which works almost miraculously well at scale. There, you get to pass one  or more template lambdas to a sum type and the library simply iterates all possibilities and picks the first that compiles. Experience with sumtype suggests that's a really solid idiom for D.

There are certain advantages to this strategy as implemented here compared to the existing code:

* It's easy to understand and maintain - each pattern is in the clear. This is a very important aspect! It's very difficult to make heads and tails of the code otherwise. Once you see what patterns the library is looking for, it's very easy to get into it.
* It's much more terse than the baseline - this PR deletes 166 lines, many of which I think are attributable to this change.
* There's good precedent with sumtype